File: Очереди с приоритетами/second.py
def HeapSort(arr):
  for start in range((len(arr)) // 2, -1, -1):
    siftDown(arr, start, len(arr) - 1) 

  print(arr)
  for end in range(len(arr) - 1, 0, -1): 
    arr[end], arr[0] = arr[0], arr[end]
    siftDown(arr, 0, end - 1)
  return arr

def siftDown(arr, start, end):
  root = start 

  while True:
    child = root * 2 + 1
    if child > end: break 

    if child + 1 <= end and arr[child] < arr[child + 1]:
      child += 1

    if arr[root] < arr[child]:
      arr[root], arr[child] = arr[child], arr[root]
      root = child
    else:
      break

print([12, 11, 13, 5, 6, 7, 10, 15])
print(HeapSort([12, 11, 13, 5, 6, 7, 10, 15]))

File: Очереди с приоритетами/second.cpp
#include <iostream>
#include <vector>

using namespace std;

void vecOut(vector<int> vec){
  cout << "[";
  for(int i = 0; i < vec.size(); i++){
    if(i != vec.size() - 1){
      cout << vec[i] << ", ";
    }else{
      cout << vec[i];
    }
  }
  cout << "]" << endl;
}

void siftDown(vector<int> &vec, int start, int end){
  int root = start;

  while(true){
    int child = root * 2 + 1;
    if(child > end) break;

    if (child + 1 <= end && vec[child] < vec[child + 1]){
      child += 1;
    }

    if (vec[root] < vec[child]){
      swap(vec[root], vec[child]);
      root = child;
    }else{
      break;
    }
  }
}

vector<int> HeapSort (vector<int> vec){
  for(int start = vec.size() / 2; start > -1; start -= 1){
    siftDown(vec, start, vec.size() - 1);
  }

  vecOut(vec);

  for(int end = vec.size() - 1; end > 0; end--){
    swap(vec[end], vec[0]);
    siftDown(vec, 0, end - 1);
  }

  return vec;
}

int main() {
  vector<int> vec = {12, 11, 13, 5, 6, 7, 10, 15};
  vecOut(vec);

  vector<int> newVec = HeapSort(vec);
  vecOut(newVec);
}
